doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,314 @@
+other Source does today. The reason we chose `environmentd` to primarily handle webhook requests is
+because it is already setup to handle _incoming_ requests from an external network. All of our
+other sources are pull based, where we send a request and ask for data. Whereas webhooks are push
+based, we receive a request, when there is new data available. Exposing a `clusterd` outside of our
+network, maybe isn't even something we want to do (e.g. in the future maybe there is a reverse
+proxy that talks to `clusterd`), and would require significant lift from the Cloud Team.  We also
+have a separate project in the works that will allow `environmentd` to scale horizontally, which
+will mitigate these concerns.
+
+### Tables vs Sources vs What we Chose
+
+An alternative approach is to use a `TABLE` for these new sources, and "translate" any incoming
+requests to `INSERT` statements. There are two reasons we did not chose this approach:
+
+1. In the future to support features like validation or custom mapping of the request body, we'll
+need a storage owned Source so we can run this logic. The migration path from a `TABLE` to this new
+kind of Source isn't entirely clear.
+2. `INSERT`s into a table are linearizable, whereas requests from a webhook should not be. We can
+update `TABLE`s to make `INSERT`s optionally linearizable, but this work probably isn't worth it
+given we also have reason 1.
+
+The second alternative is to build this new kind of storage owned Source, which gets installed on a
+cluster and can do more complex operations everytime a request is received. We plan to build
+something like this in the future, but at the moment this would require a large amount of work,
+and we first we want to validate that webhooks are a feature users actually want.
+
+## Future
+
+There is plenty of future work that can be done for Webhooks, but by limiting our scope and
+shipping quickly, we can collect feedback from users to help inform our priorities.
+
+1. A `clusterd` primarily handles requests. This would improve our ability to scale, e.g. 100s of
+QPS per source, and allow us to build the more complex features listed below.
+2. `TRANSFORM USING` a given request. Being able to map a JSON (or Avro, protobuf, etc.) body to
+a relation at the time of receiving a request, would be a win for the user as it's more ergonomic,
+and a win for us to reduce the amount of storage used.
+3. Subsources and batching. Being able map a single event to multiple independent sources, could
+be a big win for the ergonomics of the feature.
+4. User configuration of request limits via `ALTER <source>`, i.e. rate and size.
+5. Consider using some "external ID" that users can use to reference Webhook sources, e.g.
+   `/api/webhook/abcdefg`. This would allow users to drop and re-create sources without needing to
+   update external systems.
+   See <https://github.com/MaterializeInc/materialize/pull/20002#discussion_r1234726035> for more
```

Comment:
So glad you linked the discussion here. Thank you!

Reply:
No problem!